src/core/dep-graph.ts

```diff
@@ -175,6 +191,70 @@ class DepGraphImpl implements types.DepGraphInternal {
     };
   }
 
+  private nodeEquals(
+    graphA: types.DepGraphInternal,
+    nodeIdA: string,
+    graphB: types.DepGraphInternal,
+    nodeIdB: string,
+    compareRoot: boolean,
+    traversedNodesA = new Set<string>(),
+    traversedNodesB = new Set<string>(),
+  ): boolean {
+    // Skip root nodes comparision if needed.
+    if (compareRoot || (nodeIdA !== graphA.rootNodeId && nodeIdB !== graphB.rootNodeId)) {
+      const pkgA = graphA.getNodePkg(nodeIdA);
+      const pkgB = graphB.getNodePkg(nodeIdB);
+
+      // Compare PkgInfo (name and version).
+      if (!_.isEqual(pkgA, pkgB)) {
+        return false;
+      }
+
+      const infoA = graphA.getNode(nodeIdA);
+      const infoB = graphB.getNode(nodeIdB);
+
+      // Compare NodeInfo (VersionProvenance and labels).
+      if (!_.isEqual(infoA, infoB)) {
+        return false;
+      }
+    }
+
+    let depsA = graphA.getNodeDepsNodeIds(nodeIdA);
+    let depsB = graphB.getNodeDepsNodeIds(nodeIdA);
+
+    // Number of dependencies should be the same.
+    if (depsA.length !== depsB.length) {
+      return false;
+    }
+
+    // Sort dependencies by name@version string.
+    const sortFn = (graph: types.DepGraphInternal) => (idA: string, idB: string) => {
+      const pkgA = graph.getNodePkg(idA);
+      const pkgB = graph.getNodePkg(idB);
+      return DepGraphImpl.getPkgId(pkgA).localeCompare(DepGraphImpl.getPkgId(pkgB));
+    };
+
+    depsA = depsA.sort(sortFn(graphA));
+    depsB = depsB.sort(sortFn(graphB));
+
+    // Compare Each dependency recursively.
+    for (let i = 0; i < depsA.length; i++) {
+      // Prevent cycles.
+      if (traversedNodesA.has(depsA[i]) && traversedNodesB.has(depsB[i])) {
```

Comment:
@Kirill89 maybe I miss something, but shouldn't we check that the `depsA[i]->package` equals `depsB[i]->package` ? what if both have cycles but with different packages?

Reply:
We check it in the recursive call.

Reply:
Agree after discussion.